src/lib/VirtualScroller.js

```diff
@@ -0,0 +1,183 @@
+        this.contentEl.style.height = `${this.totalItems * this.itemHeight + this.marginTop + this.marginBottom}px`;
+
+        this.containerEl.appendChild(this.contentEl);
+        this.container.appendChild(this.containerEl);
+
+        this.renderItems();
+
+        this.bindDOMListeners();
+    }
+
+    /**
+     * Binds DOM listeners
+     *
+     * @return {void}
+     */
+    bindDOMListeners() {
+        this.containerEl.addEventListener('scroll', this.onScrollHandler);
+    }
+
+    /**
+     * Unbinds DOM listeners
+     *
+     * @return {void}
+     */
+    unbindDOMListeners() {
+        if (this.containerEl) {
+            this.containerEl.removeEventListener('scroll', this.onScrollHandler);
+        }
+    }
+
+    /**
+     * Handler for 'scroll' event
+     *
+     * @param {Event} e - The scroll event
+     * @return {void}
+     */
+    onScrollHandler(e) {
+        const { scrollTop } = e.target;
+
+        if (Math.abs(scrollTop - this.previousScrollTop) > this.maxBufferHeight) {
+            // The first item to be re-rendered will be a totalViewItems height up from the
+            // item at the current location
+            const firstIndex = Math.floor(scrollTop / this.itemHeight) - this.totalViewItems;
+            this.renderItems(firstIndex < 0 ? 0 : firstIndex);
+
+            this.previousScrollTop = scrollTop;
+        }
+    }
+
+    /**
+     * Render a set of items, starting from the offset index
+     *
+     * @param {number} offset  - The offset to start rendering items
+     * @return {void}
+     */
+    renderItems(offset = 0) {
+        let count = this.maxRenderedItems;
+        // If the default count of items to render exceeds the totalItems count
+        // then just render the difference
+        if (count + offset > this.totalItems) {
+            count = this.totalItems - offset;
+        }
+
+        let numItemsRendered = 0;
+        const fragment = document.createDocumentFragment();
+
+        while (numItemsRendered < count) {
+            const rowEl = this.renderItem(offset + numItemsRendered, this.itemHeight);
+            fragment.appendChild(rowEl);
+            numItemsRendered += 1;
+        }
+
+        this.contentEl.innerHTML = '';
+        this.contentEl.appendChild(fragment);
+    }
+
+    /**
+     * Render a single item
+     *
+     * @param {number} rowIndex - The index of the item to be rendered
+     * @return {HTMLElement} The newly created row item
+     */
+    renderItem(rowIndex) {
+        const rowEl = document.createElement('div');
+        const topPosition = this.itemHeight * rowIndex + this.marginTop;
+
+        let renderedThumbnail;
+        try {
+            renderedThumbnail = this.renderItemFn.call(this, rowIndex);
+        } catch (err) {
+            // eslint-disable-next-line
+            console.error(`Error rendering thumbnail - ${err}`);
+        }
+
+        rowEl.style.top = `${topPosition}px`;
```

Comment:
Consider moving the row "component" to its own file. Also, if you want to get fancy, you could use template strings for these elements; it'd give you a JSX-ish syntax.